workflow/executor/agent.go

```diff
@@ -31,79 +31,135 @@ type AgentExecutor struct {
 	WorkflowInterface workflow.Interface
 	RESTClient        rest.Interface
 	Namespace         string
-	CompleteTask      map[string]struct{}
+	consideredTasks   map[string]bool
+}
+
+func NewAgentExecutor(clientSet kubernetes.Interface, restClient rest.Interface, config *rest.Config, namespace, workflowName string) *AgentExecutor {
+	return &AgentExecutor{
+		ClientSet:         clientSet,
+		RESTClient:        restClient,
+		Namespace:         namespace,
+		WorkflowName:      workflowName,
+		WorkflowInterface: workflow.NewForConfigOrDie(config),
+		consideredTasks:   make(map[string]bool),
+	}
+}
+
+type agentTask struct {
+	NodeId   string
+	Template wfv1.Template
+}
+
+type patchResponse struct {
+	NodeId string
+	Patch  []byte
 }
 
 func (ae *AgentExecutor) Agent(ctx context.Context) error {
 	defer runtimeutil.HandleCrash(runtimeutil.PanicHandlers...)
 
+	log.Info("Starting Agent")
+
+	agentTaskQueue := make(chan agentTask)
+	patchResponseQueue := make(chan patchResponse)
 	taskSetInterface := ae.WorkflowInterface.ArgoprojV1alpha1().WorkflowTaskSets(ae.Namespace)
+
+	go ae.patchWorker(ctx, taskSetInterface, patchResponseQueue)
+	for i := 0; i < 8; i++ {
```

Comment:
Default 16? Configurable?

Reply:
configurable is good. can we do it through Env and the controller will set ENV when creating POD?